File: core0/us/sdk/wallet/java/test/us/gov/crypto/test/symmetric_encryption.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.gov.crypto.test;

import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.KeyPair;
import java.security.GeneralSecurityException;
import java.util.Arrays;
import us.pair;
import us.ko;

public class symmetric_encryption {

    public static void log(final String line) { //--strip
       System.out.println("symmetric_encryption tests: "+line); //--strip
    } //--strip

    KeyPair a;
    KeyPair b;
    KeyPair c;

    public boolean main() {

        assert us.gov.crypto.symmetric_encryption.test();

        log("Creating keys"); //--strip
        a = us.gov.crypto.ec.instance.generate_keypair();
        if (a==null) return false;
        b = us.gov.crypto.ec.instance.generate_keypair();
        if (b==null) return false;
        c = us.gov.crypto.ec.instance.generate_keypair();
        if (c==null) return false;
        PrivateKey priv_a = a.getPrivate();
        PublicKey pub_a = a.getPublic();
        PrivateKey priv_b = b.getPrivate();
        PublicKey pub_b = b.getPublic();
        PrivateKey priv_c = c.getPrivate();
        PublicKey pub_c = c.getPublic();
        log("tests enc/dec"); //--strip
        //test that encryption then decryption will retrieve original plaintext.
        assert(test_encrypt_decrypt("encrypt this"                                                                                                                        ));
        assert(test_encrypt_decrypt(""                                                                                                                                    ));
        assert(test_encrypt_decrypt(",./;'#[]-=123456DFJLSKDFJERUEIUR  \n rtr"                                                                                          ));
        assert(test_encrypt_decrypt("0"                                                                                                                                    ));
        assert(test_encrypt_decrypt("𝕋𝕙𝕖 𝕢𝕦𝕚𝕔𝕜 𝕓𝕣𝕠𝕨𝕟 𝕗𝕠𝕩 𝕛𝕦𝕞𝕡𝕤 𝕠𝕧𝕖𝕣 𝕥𝕙𝕖 𝕝𝕒𝕫𝕪 𝕕𝕠𝕘"                                                                                     ));
        assert(test_encrypt_decrypt("パーティーへ行かないか"                                                                                                                 ));
        assert(test_encrypt_decrypt("˙ɐnbᴉlɐ ɐuƃɐɯ ǝɹolop ʇǝ ǝɹoqɐl ʇn ʇunpᴉpᴉɔuᴉ ɹodɯǝʇ poɯsnᴉǝ op pǝs 'ʇᴉlǝ ƃuᴉɔsᴉdᴉpɐ ɹnʇǝʇɔǝsuoɔ 'ʇǝɯɐ ʇᴉs ɹolop ɯnsdᴉ ɯǝɹo˥00˙Ɩ$-" ));
        assert(test_encrypt_decrypt("null"                                                                                                                                ));
        assert(test_encrypt_decrypt("-$1.00"                                                                                                                            ));
        log("tests enc 1"); //--strip

        //test that same plaintext is encrypted to different ciphertexts.
        assert(test_encrypt_multiple("The ciphertext should be different although the plaintext is the same"));

        //test that ciphertext is decrypted to the same plaintext.
        assert(test_decrypt_multiple("The plaintext should be the same each time we decrypt"));

        //test that message can't be decoded with the wrong key.
        assert(!test_encrypt_decrypt_keys("test that message can't be decoded with the wrong key", priv_a, pub_b, priv_c, pub_a));

        //decrypting invalid ciphertext should return an empty vector.
        assert(test_decrypt_nulls(""));
        assert(test_decrypt_nulls("this string hasn't been encrypted so decryption will fail"));
        return true;

    }

    public boolean test_encrypt_decrypt(String plaintext_string)  {
        KeyPair a = us.gov.crypto.ec.instance.generate_keypair();
        if (a==null) return false;
        KeyPair b = us.gov.crypto.ec.instance.generate_keypair();
        if (b==null) return false;
        PrivateKey priv_a = a.getPrivate();
        PublicKey pub_a = a.getPublic();
        PrivateKey priv_b = b.getPrivate();
        PublicKey pub_b = b.getPublic();
        log("test_encrypt_decrypt_keys"); //--strip
        return test_encrypt_decrypt_keys(plaintext_string, priv_a, pub_b, priv_b, pub_a);
    }

    public boolean test_encrypt_decrypt_keys(String plaintext_string, PrivateKey priv_a, PublicKey pub_b, PrivateKey priv_b, PublicKey pub_a) {
        log("plaintext_string: "+plaintext_string); //--strip
        byte[] plaintext = plaintext_string.getBytes();
        us.gov.crypto.symmetric_encryption se_a= new us.gov.crypto.symmetric_encryption();
        ko r=se_a.init(priv_a,pub_b);
        if (ko.is_ko(r)) {
            log("A "+r.msg); //--strip
            assert false;
        }
        pair<ko, byte[]> ciphertext = se_a.encrypt(plaintext);
        us.gov.crypto.symmetric_encryption se_b= new us.gov.crypto.symmetric_encryption();
        ko r2=se_b.init(priv_b, pub_a);
        if (ko.is_ko(r2)) {
            log("B "+r2.msg); //--strip
            assert false;
        }

        {
            pair<ko, byte[]> decryptedtext = se_b.decrypt(ciphertext.second);
            if (ko.is_ko(decryptedtext.first)) return false;
            assert Arrays.equals(plaintext, decryptedtext.second);
        }
        {
            //System.out.println("E2");
            byte[] buf=new byte[ciphertext.second.length+500];
            for (int i=0; i<ciphertext.second.length; ++i) {
                buf[200+i]=ciphertext.second[i];
            }
            pair<ko, byte[]> decryptedtext = se_b.decrypt(buf, 200, ciphertext.second.length);
            //System.out.println("E2-0");
            if (ko.is_ko(decryptedtext.first)) return false;
            //System.out.println("E2-1");
            assert Arrays.equals(plaintext, decryptedtext.second);
            //System.out.println("E2-2");
        }
        return true;
    }

    public boolean test_encrypt_multiple(String plaintext_string) {

        KeyPair a = us.gov.crypto.ec.instance.generate_keypair();
        if (a==null) return false;
        KeyPair b = us.gov.crypto.ec.instance.generate_keypair();
        if (b==null) return false;

        PrivateKey priv_a = a.getPrivate();
        PublicKey pub_a = a.getPublic();
        PrivateKey priv_b = b.getPrivate();
        PublicKey pub_b = b.getPublic();

        byte[] plaintext = plaintext_string.getBytes();

        us.gov.crypto.symmetric_encryption se_a = new us.gov.crypto.symmetric_encryption();
        ko r=se_a.init(priv_a, pub_b);
        if (ko.is_ko(r)) return false;
        pair<ko, byte[]> ciphertext_1 = se_a.encrypt(plaintext);
        if (ko.is_ko(ciphertext_1.first)) return false;
        pair<ko, byte[]> ciphertext_2 = se_a.encrypt(plaintext);
        if (ko.is_ko(ciphertext_2.first)) return false;
        return !Arrays.equals(ciphertext_1.second, ciphertext_2.second);
    }

    public boolean test_decrypt_nulls(String ciphertext_string) {

        KeyPair a = us.gov.crypto.ec.instance.generate_keypair();
        if (a==null) return false;
        KeyPair b = us.gov.crypto.ec.instance.generate_keypair();
        if (b==null) return false;
//System.out.println("A0");
        PrivateKey priv_a = a.getPrivate();
        PublicKey pub_a = a.getPublic();
        PrivateKey priv_b = b.getPrivate();
        PublicKey pub_b = b.getPublic();

        byte[] ciphertext = ciphertext_string.getBytes();
        us.gov.crypto.symmetric_encryption se = new us.gov.crypto.symmetric_encryption();
        ko r=se.init(priv_a, pub_b);
//System.out.println("A1");
        if (ko.is_ko(r)) return false;
//System.out.println("A2 "+ ciphertext.length);
        pair<ko, byte[]> decryptedtext = se.decrypt(ciphertext);
//System.out.println("A3");
        if (ko.is_ok(decryptedtext.first)) return false;
        if (ciphertext.length<us.gov.crypto.symmetric_encryption.iv_size) {
 //           System.out.println("A3<");
            assert decryptedtext.first==us.gov.crypto.symmetric_encryption.KO_89601;
        }
        else {
 //           System.out.println("A3>");
            assert decryptedtext.first==us.gov.crypto.symmetric_encryption.KO_40751;
        }
//System.out.println("A4");
        return true;
//        return Arrays.equals(decryptedtext.second,new byte[0]);
    }

    public boolean test_decrypt_multiple(String plaintext_string) {

        KeyPair a = us.gov.crypto.ec.instance.generate_keypair();
        if (a==null) return false;
        KeyPair b = us.gov.crypto.ec.instance.generate_keypair();
        if (b==null) return false;

        PrivateKey priv_a = a.getPrivate();
        PublicKey pub_a = a.getPublic();
        PrivateKey priv_b = b.getPrivate();
        PublicKey pub_b = b.getPublic();

        byte[] plaintext = plaintext_string.getBytes();

        us.gov.crypto.symmetric_encryption se_a = new us.gov.crypto.symmetric_encryption();
        ko r=se_a.init(priv_a, pub_b);
        if (ko.is_ko(r)) return false;

        pair<ko, byte[]> ciphertext = se_a.encrypt(plaintext);
        if (ko.is_ko(ciphertext.first)) return false;
        us.gov.crypto.symmetric_encryption se_b = new us.gov.crypto.symmetric_encryption();
        ko r2=se_b.init(priv_b, pub_a);
        if (ko.is_ko(r2)) return false;
        pair<ko, byte[]> decryptedtext1 = se_b.decrypt(ciphertext.second);
        pair<ko, byte[]> decryptedtext2 = se_b.decrypt(ciphertext.second);
        if (ko.is_ko(decryptedtext1.first)) return false;
        if (ko.is_ko(decryptedtext2.first)) return false;
        return Arrays.equals(decryptedtext1.second, decryptedtext2.second);
    }

}

File: core0/us/gov/crypto/symmetric_encryption.cpp
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "symmetric_encryption.h"

#include <iterator>

#include <us/gov/likely.h>

#include <secp256k1_ecdh.h>
#include <crypto++/cryptlib.h>
#include <crypto++/filters.h>

#define loglevel "gov/crypto"
#define logclass "symmetric_encryption"
#include "logs.inc"

using namespace us::gov::crypto;
using c = us::gov::crypto::symmetric_encryption;
using us::ko;
using CryptoPP::BufferedTransformation;
using CryptoPP::AuthenticatedSymmetricCipher;
using CryptoPP::Redirector;
using CryptoPP::StringSink;
using CryptoPP::StringSource;
using CryptoPP::ArraySink;
using CryptoPP::ArraySource;
using CryptoPP::AuthenticatedEncryptionFilter;
using CryptoPP::AuthenticatedDecryptionFilter;


ko c::init(const keys::priv_t& sk, const keys::pub_t& pub_other) {
    log("init. pub_other:", pub_other);
    ko r = ec::instance.generate_shared_key(key_, sizeof(key_), sk, pub_other);
    if (unlikely(is_ko(r))) {
        log(r);
        return r;
    }
    return ok;
}

namespace {
    void log_ratio(size_t e, size_t d) {   //--strip
        log("IV sz ", c::iv_size,              //--strip
            "scrambled msg sz", e,         //--strip
            "enc/dec ratio(", d, ") = ", (double)e / (double)d);    //--strip
    }                                         //--strip
}

ko c::encrypt(const vector<unsigned char>& cleartext, vector<unsigned char>& ciphertext, size_t ciphertext_offset) {
    log("encrypt block of", cleartext.size(), "bytes.");
    size_t estimated_sz = ciphertext_offset + iv_size + cleartext.size() + AES::BLOCKSIZE;
    log("estimated size", estimated_sz);
    ciphertext.resize(estimated_sz);
    try {
        unsigned char iv_[iv_size];
        prng_.GenerateBlock(iv_, iv_size);
        GCM<AES>::Encryption enc;
        enc.SetKeyWithIV(key_, key_size, iv_, iv_size);
        ArraySink sink(ciphertext.data() + ciphertext_offset, ciphertext.size() - ciphertext_offset);
        ArraySource(cleartext.data(), cleartext.size(), true, new AuthenticatedEncryptionFilter(enc, new Redirector(sink), false, tag_size));
        size_t final_sz = ciphertext_offset + sink.TotalPutLength() + iv_size;
        log("final size", final_sz);
        ciphertext.resize(final_sz);
        assert(estimated_sz >= final_sz); //if fails here adjust encmsg_max_sz
        log_ratio(ciphertext.size() - ciphertext_offset, cleartext.size()); //--strip
        memcpy(ciphertext.data() + ciphertext_offset + sink.TotalPutLength(), iv_, iv_size);
        log("total encrypted payload", ciphertext.size() - ciphertext_offset, "bytes.");
    }
    catch(CryptoPP::InvalidArgument& e) {
        auto r = "KO 30921 Invalid argument in encryption.";
        log(r, e.what());
        return r;
    }
    return ok;
}

ko c::decrypt(const vector<unsigned char>& ciphertext, vector<unsigned char>& cleartext) {
    return decrypt(ciphertext.data(), ciphertext.size(), cleartext);
}

ko c::decrypt(const unsigned char* ciphertext, size_t ciphertext_sz, vector<unsigned char>& cleartext) {
    if (unlikely(ciphertext_sz < iv_size)) {
        auto r = "KO 44031 Message too small to decrypt.";
        log(r);
        return r;
    }
    size_t sz_msg = ciphertext_sz - iv_size;
    log("decrypt block of ", ciphertext_sz, "bytes. msg sz=", sz_msg, "+ IV size", iv_size);
    cleartext.resize(ciphertext_sz);
    try {
        GCM<AES>::Decryption dec;
        dec.SetKeyWithIV(key_, key_size, &ciphertext[sz_msg], iv_size);
        ArraySink sink(cleartext.data(), cleartext.size());
        AuthenticatedDecryptionFilter filter(dec, new Redirector(sink), AuthenticatedDecryptionFilter::DEFAULT_FLAGS, tag_size);
        ArraySource(ciphertext, sz_msg, true, new Redirector(filter));
        cleartext.resize(sink.TotalPutLength());
        log_ratio(ciphertext_sz, cleartext.size()); //--strip
    }
    catch (const CryptoPP::HashVerificationFilter::HashVerificationFailed& e) {
        ko r = "KO 44032 CryptoPP::HashVerificationFilter::HashVerificationFailed.";
        log(r, e.what());
        return r;
    }
    catch (const CryptoPP::InvalidArgument& e) {
        auto r = "KO 44033 Invalid decrypting argument.";
        log(r, e.what());
        return r;
    }
    catch (const CryptoPP::Exception& e) {
        auto r = "KO 44034 Decryption was unsuccessful.";
        log(r, e.what());
        return r;
    }
    return ok;
}

